recipe-server/client/actions/preference-experiment/index.js

```diff
@@ -31,38 +25,35 @@ export default class PreferenceExperimentAction extends Action {
     // Exit early if we're on an incompatible client.
     if (experiments === undefined) {
       this.normandy.log('Client does not support preference experiments, aborting.', 'info');
-      return Promise.resolve();
+      return;
     }
 
     seenExperimentNames.push(slug);
-    return experiments.has(slug).then(hasSlug => {
-      // If the experiment doesn't exist yet, enroll!
-      if (!hasSlug) {
-        return this.chooseBranch(branches).then(branch =>
-          experiments.start({
-            name: slug,
-            branch: branch.slug,
-            preferenceName,
-            preferenceValue: branch.value,
-            preferenceBranchType,
-            preferenceType,
-          })
-        );
-      }
-
-      // If the experiment exists, and isn't expired, bump the lastSeen date.
-      return experiments.get(slug).then(experiment => {
-        if (experiment.expired) {
-          this.normandy.log(`Experiment ${slug} has expired, aborting.`, 'debug');
-          return true;
-        }
 
-        return experiments.markLastSeen(slug);
+    // If the experiment doesn't exist yet, enroll!
+    const hasSlug = await experiments.has(slug);
+    if (!hasSlug) {
+      const branch = await this.chooseBranch(branches);
+      await experiments.start({
+        name: slug,
+        branch: branch.slug,
+        preferenceName,
+        preferenceValue: branch.value,
+        preferenceBranchType,
+        preferenceType,
       });
-    });
+    } else {
+      // If the experiment exists, and isn't expired, bump the lastSeen date.
+      const experiment = await experiments.get(slug);
+      if (experiment.expired) {
```

Comment:
It might be worth it! I have opinions about this that I haven't really articulated before.

I run into situations like this a lot (a complex conditional that could split into a few standalone functions) and what I've found is that moving parts of the conditional into standalone functions is less comprehensible most of the time. The reason is that the purpose of the function containing the larger conditional is generally to dispatch to one or more smaller operations. 

Here, `execute`'s purpose can be viewed as "given the client state, what should we do about this preference experiment?". Moving part of the conditional into another function makes it harder to understand how this dispatch will end up, because now you need to look at two functions to determine what choice will be made. It's like a state machine where you can only see half of the nodes at a given time and need to swap between the two ends.

What I usually do (and tried to do here), is to factor out the action that you end up taking into standalone functions, but keep the conditional checking in the main function. Sometimes (and this might be an appropriate place) it's also worth it to factor out the condition into a standalone function while keeping the conditional itself in the main function. Like this:

```js
} else if (!(await this.isExpired(slug))) {
  await experiments.markLastSeen(slug);
} else {
  this.normandy.log(`Experiment ${slug} has expired, aborting.`, 'debug');
}
```

One less level of indentation makes it more readable, and it's pretty easy to name the function such that the conditional is understandable. I was kinda iffy on the `!` in front of the await though; I've never liked `!(await blah)` vs `!await blah`. I didn't give it much more thought and just went with what's here.

For the same reasons, I also generally prefer having a full `if` statement with an `else` even if the `if` has a return inside, but our linter currently doesn't like that so I don't do it as often.

What do you think? Should we go with the alternative, what's there, or do you still like your suggestion?